### PR TITLE
Add synthetic current backend and correct SITL current computation

### PR DESCRIPTION
**Pull request by Hwurzburg**
replaces #22117 per dev call

### CODE REVIEW

**Comment by tridge** [comment]
in tfs_validate_current1() need to use math.round() to round both values to closest 0.1A (deci-amp)

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -11880,7 +11880,7 @@ def tfs_validate_current1(self, value):
         if batt is None:
             raise NotAchievedException("Did not get BATTERY_STATUS message")
         battery_status_current_a = batt.current_battery * 0.01 # cA -> A
-        self.progress("BATTERY_STATUS current==%f frsky==%f" % (battery_status_current_a, current_a))
+        self.progress("BATTERY_STATUS current==%f frsky==%f" % (math.round(battery_status_current_a, 1), math.round(current_a, 1)) # noqa: E501
         if self.compare_number_percent(battery_status_current_a, current_a, 10):
```
Comment (peterbarker) [inline Tools/autotest/common.py:11884]: Rounding needs to happen on this line, not the previous one

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -11880,7 +11880,7 @@ def tfs_validate_current1(self, value):
         if batt is None:
             raise NotAchievedException("Did not get BATTERY_STATUS message")
         battery_status_current_a = batt.current_battery * 0.01 # cA -> A
-        self.progress("BATTERY_STATUS current==%f frsky==%f" % (battery_status_current_a, current_a))
+        self.progress("BATTERY_STATUS current==%f frsky==%f" % (math.round(battery_status_current_a, 1), math.round(current_a, 1)) # noqa: E501
```
Comment (peterbarker) [inline Tools/autotest/common.py:11883]:
```suggestion
        self.progress("BATTERY_STATUS current==%f frsky==%f" % (math.round(battery_status_current_a, 1), math.round(current_a, 1)))
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: Tools/autotest/common.py (hunk 1 above)
Comment (Hwurzburg) [reply Tools/autotest/common.py]: done

File: Tools/autotest/common.py (hunk 2 above)
Comment (Hwurzburg) [reply Tools/autotest/common.py]: now I understand...was going by Tridge's line number

**Comment by Hwurzburg** [comment]
@tridge apparently our python "math" has no round function

**Comment by Hwurzburg** [comment]
Tested on Vehicle
BAT 0 Curr
first ran full throttle on another log and set params : max volt 12.6 max current 13.6 then ran again
resting voltage the same (should be if the currents match and using the same voltage sensor)
looks good to me

**Review by tridge — APPROVED** [review]

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 3]
```diff
@@ -44,6 +44,10 @@
 #define AP_BATTMON_FUELLEVEL_ANALOG_ENABLE (BOARD_FLASH_SIZE > 1024)
 #endif
 
+#ifndef AP_BATTMON_SYNTHETIC_CURRENT_ENABLED
+#define AP_BATTMON_SYNTHETIC_CURRENT_ENABLED  !defined(HAL_BUILD_AP_PERIPH)
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor.h:48]: Please put this into `chibios_hwdef.py` instead.

```
diff --git a/libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py b/libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py
index 7703d6079a3..906454011f0 100644
--- a/libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py
+++ b/libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py
@@ -2942,6 +2942,11 @@ def add_apperiph_defaults(f):
 #ifndef AP_COMPASS_DIAGONALS_ENABLED
 #define AP_COMPASS_DIAGONALS_ENABLED 0
 #endif
+
+// disable various battery monitor backends:
+#ifndef AP_BATTMON_SYNTHETIC_CURRENT_ENABLED
+#define AP_BATTMON_SYNTHETIC_CURRENT_ENABLED 0
+#endif
 ''')
 
 def add_bootloader_defaults(f):
```

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 4]
```diff
@@ -113,6 +119,7 @@ class AP_BattMonitor
         LTC2946                    = 22,
         Torqeedo                   = 23,
         FuelLevel_Analog           = 24,
+        Analog_volt_synth_current  = 25,
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor.h:122]:
```suggestion
        Analog_Volt_Synthetic_Current  = 25,
```

File: libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    // read current
+    // calculate time since last current read
+    const uint32_t tnow = AP_HAL::micros();
+    const uint32_t dt_us = tnow - _state.last_time_micros;
+
+    // this copes with changing the pin at runtime
+    _state.healthy &= _curr_pin_analog_source->set_pin(_curr_pin);
+
+    // read current
+    _state.current_amps = ((_state.voltage/_max_voltage)*sq(gcs().get_hud_throttle()) * 0.0001 * _curr_amp_per_volt) + _curr_amp_offset ;
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp:63]: I'm really not happy getting this from the GCS library - it's a pretty nasty layering violation.

When `get_hud_throttle` was created we didn't have `AP_Vehicle` which the other vehicles inherit from.

We need a method on `AP_Vehicle`,  `float get_throttle() const`:

**Review by peterbarker — CHANGES_REQUESTED** [review]
Is this going to be wildly wrong in the face of separate-thrust-and-lift quadplanes?

File: libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp (hunk 5 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp:63]: @IamPete1 suggested we can use SRV_Channels::get_output_scaled(SRV_Channel::k_throttle) which will work for fixed wing and for fwd motor on quadplanes. I think that is a good solution

Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp:63]: Swipe from this?
```
    nav_scripting.throttle_pct = SRV_Channels::get_output_scaled(SRV_Channel::k_throttle);
```

File: libraries/AP_BattMonitor/AP_BattMonitor.h (hunk 3 above)
Comment (Hwurzburg) [reply libraries/AP_BattMonitor/AP_BattMonitor.h:48]: done

File: libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp (hunk 5 above)
Comment (Hwurzburg) [reply libraries/AP_BattMonitor/AP_BattMonitor_Synthetic_Current.cpp:63]: done

File: libraries/AP_BattMonitor/AP_BattMonitor.h (hunk 4 above)
Comment (Hwurzburg) [reply libraries/AP_BattMonitor/AP_BattMonitor.h:122]: done

**Review by peterbarker — APPROVED** [review]
LGTM

**Review by tridge — APPROVED** [review]

**Pull request merged**